compiler_gym/util/minimize_trajectory.py

```diff
@@ -0,0 +1,317 @@
+                "🟢 Hypothesis holds with %s of %s actions randomly removed, continuing",
+                num_to_remove,
+                len(actions),
+            )
+            actions = candidate_actions
+            discard_ratio = init_discard_ratio
+            yield env
+        else:
+            logging.info(
+                "🔴 Hypothesis does not hold with %s of %s actions randomly removed, rolling back",
+                num_to_remove,
+                len(actions),
+            )
+            discard_ratio *= discard_ratio_decay
+            if num_to_remove == 1:
+                logging.info(
+                    "Terminating random minimization after failing with only a single action removed"
+                )
+                break
+
+    if not apply_and_test(actions):
+        raise MinimizationError("Post-minimization sanity check failed!")
+
+    yield env
+
+
+def minimize_trajectory_iteratively(
+    env: CompilerEnv,
+    hypothesis: Hypothesis = environment_validation_fails,
+    flakiness: int = 1,
+) -> Iterable[CompilerEnv]:
+    """Minimize a trajectory by remove actions, one at a time, until a minimal
+    trajectory is reached.
+
+    Performs up to O(n * n / 2) evaluation rounds, where n is the length of the
+    trajectory.
+
+    :param env: An environment whose action trajectory should be minimized.
+    :param hypothesis: The hypothesis that is used to determine if a trajectory
+        is valid. A callback that accepts as argument the :code:`env:`
+        instance and returns true if the hypothesis holds, else false. The
+        hypothesis must hold on the initial trajectory.
+    :param flakiness: The maximum number of times the hypothesis is repeated
+        to check if it holds.
+    :returns: A generator that yields the input environment every time the
+        trajectory is successfully reduced.
+    :raises MinimizationError: If the minimization fails, or if the hypothesis
+        does not hold on the initial trajectory.
+    """
+
+    def apply_and_test(actions):
+        return _apply_and_test(env, actions, hypothesis, flakiness)
+
+    all_actions = env.actions.copy()
+    init_num_actions = len(all_actions)
+    if not all_actions:
+        raise ValueError("No actions in the given state.")
+
+    if not apply_and_test(all_actions):
+        raise ValueError(
+            "Hypothesis failed on the initial state! The hypothesis must hold for the first state."
+        )
+
+    pass_num = 0
+    actions_removed = 0
+    action_has_been_pruned = True
+    # Outer loop. Repeat iterative reduction until no change is made.
+    while action_has_been_pruned and len(all_actions) > 1:
+        pass_num += 1
+        action_has_been_pruned = False
+        action_mask = [True] * len(all_actions)
+        logging.info("Minimization pass on sequence of %d actions", len(all_actions))
+
+        # Inner loop. Go through every action and see if it can be removed.
+        for i in range(len(action_mask)):
```

Comment:
What's the mask for? Why not just delete the actions like in previous algs?

Reply:
IMO the mask is easier to read for the case where we're flicking actions on and off, but no particular reason :)